pyomo/contrib/pyros/config.py

```diff
@@ -0,0 +1,1043 @@
+        of objects of type ``self.cdatatype``.
+        """
+        if self.ctype_validator is not None:
+            self.ctype_validator(obj)
+        return list(obj.values())
+
+    def standardize_cdatatype_obj(self, obj):
+        """
+        Standardize object of type ``self.cdatatype`` to
+        ``[obj]``.
+        """
+        if self.cdatatype_validator is not None:
+            self.cdatatype_validator(obj)
+        return [obj]
+
+    def __call__(self, obj, from_iterable=None, allow_repeats=None):
+        """
+        Cast object to a flat list of Pyomo component data type
+        entries.
+
+        Parameters
+        ----------
+        obj : object
+            Object to be cast.
+        from_iterable : Iterable or None, optional
+            Iterable from which `obj` obtained, if any.
+        allow_repeats : bool or None, optional
+            True if list can contain repeated entries,
+            False otherwise.
+
+        Raises
+        ------
+        TypeError
+            If all entries in the resulting list
+            are not of type ``self.cdatatype``.
+        ValueError
+            If the resulting list contains duplicate entries.
+        """
+        if allow_repeats is None:
+            allow_repeats = self.allow_repeats
+
+        if isinstance(obj, self.ctype):
+            ans = self.standardize_ctype_obj(obj)
+        elif isinstance(obj, self.cdatatype):
+            ans = self.standardize_cdatatype_obj(obj)
+        elif isinstance(obj, Iterable) and not isinstance(obj, str):
+            ans = []
+            for item in obj:
+                ans.extend(self.__call__(item, from_iterable=obj))
+        else:
+            from_iterable_qual = (
+                f" (entry of iterable {from_iterable})"
+                if from_iterable is not None
+                else ""
+            )
+            raise TypeError(
+                f"Input object {obj!r}{from_iterable_qual} "
+                "is not of valid component type "
+                f"{self.ctype.__name__} or component data type "
+                f"{self.cdatatype.__name__}."
+            )
+
+        # check for duplicates if desired
+        if not allow_repeats and len(ans) != len(ComponentSet(ans)):
+            comp_name_list = [comp.name for comp in ans]
+            raise ValueError(
+                f"Standardized component list {comp_name_list} "
+                f"derived from input {obj} "
+                "contains duplicate entries."
+            )
+
+        return ans
+
+    def domain_name(self):
+        """Return str briefly describing domain encompassed by self."""
+        return (
+            f"{self.cdatatype.__name__}, {self.ctype.__name__}, "
+            f"or Iterable of {self.cdatatype.__name__}/{self.ctype.__name__}"
+        )
+
+
+class NotSolverResolvable(Exception):
+    """
+    Exception type for failure to cast an object to a Pyomo solver.
+    """
```

Comment:
I think I would rename this `SolverNotResolvable`.  Also, it would probably be good to make it inherit off of `PyomoException`